File: src/importer/ncnn/ops/pooling.cpp
// Tencent is pleased to support the open source community by making ncnn available.
//
// Copyright (C) 2021 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
// https://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

#include "../ncnn_importer.h"
#include "nncase/ir/node.h"
#include <cassert>
#include <nncase/ir/graph.h>
#include <nncase/ir/op_utils.h>
#include <nncase/ir/ops/binary.h>
#include <nncase/ir/ops/bitcast.h>
#include <nncase/ir/ops/constant.h>
#include <nncase/ir/ops/reduce.h>
#include <nncase/ir/ops/reduce_window2d.h>

using namespace nncase;
using namespace nncase::importer;
using namespace nncase::ir;
using namespace ncnn;

void nncase::importer::ncnn_importer::convert_op_Pooling(const Layer &layer, const ParamDict &pd, const ModelBin & /*mb*/)
{
    const int pooling_type = pd.get(0, 0);
    const int kernel_w = pd.get(1, 0);
    const int kernel_h = pd.get(11, kernel_w);
    const int stride_w = pd.get(2, 1);
    const int stride_h = pd.get(12, stride_w);
    const int pad_left = pd.get(3, 0);
    const int pad_right = pd.get(14, pad_left);
    const int pad_top = pd.get(13, pad_left);
    const int pad_bottom = pd.get(15, pad_top);
    const int global_pooling = pd.get(4, 0);
    const int pad_mode = pd.get(5, 0);
    const int avgpool_count_include_pad = pd.get(6, 0);
    const int adaptive_pooling = pd.get(7, 0);
    const int out_w = pd.get(8, 0);
    const int out_h = pd.get(18, out_w);

    const auto &op_name = layer.name;

    auto in_shape = output_tensors_.at(layer.bottoms[0])->shape();
    shape_t rshape_in_shape { 1, in_shape[0], in_shape[1], in_shape[2] };

    auto in_rshape = graph_.emplace<bitcast>(dt_float32, in_shape, rshape_in_shape);
    in_rshape->name(op_name + ".reshape(Input)");
    input_tensors_.emplace(&in_rshape->input(), layer.bottoms[0]);

    reduce_op_t reduce_type;
    float init_value;
    if (pooling_type == 0)
    {
        reduce_type = reduce_max;
        init_value = -FLT_MAX;
    }
    if (pooling_type == 1)
    {
        reduce_type = reduce_mean;
        init_value = 0.f;
    }

    value_range<float> fused_activation = value_range<float>::full();

    bool ceil_mode = false;
    ir::node *pooling_node = 0;

    if (global_pooling)
    {
        axis_t axis = { 2, 3 };
        bool keep_dims = false;
        pooling_node = graph_.emplace<reduce>(reduce_type, rshape_in_shape, axis, init_value, keep_dims);
        pooling_node->name(op_name + ".reduce(Pooling)");
    }
    if (adaptive_pooling)
    {
        const int w = in_shape[2];
        const int h = in_shape[1];
        const int kernel_extent_h = h - out_h + 1;
        const int kernel_extent_w = w - out_w + 1;

        pooling_node = graph_.emplace<reduce_window2d>(reduce_type, rshape_in_shape, init_value, kernel_extent_h, kernel_extent_w, padding { 0, 0 }, padding { 0, 0 }, 1, 1, 1, 1, fused_activation, ceil_mode, avgpool_count_include_pad);
        pooling_node->name(op_name + ".reduce_window2d(Pooling)");
    }

    padding padding_h;
    padding padding_w;
    {
        if (pad_mode == 0) // full padding
        {
            const int w = in_shape[2];
            const int h = in_shape[1];
            const int wtail = (w + pad_left + pad_right - kernel_w) % stride_w;
            const int htail = (h + pad_top + pad_bottom - kernel_h) % stride_h;

            int wtailpad = 0;
            int htailpad = 0;

            if (wtail != 0)
                wtailpad = stride_w - wtail;
            if (htail != 0)
                htailpad = stride_h - htail;

            padding_h = { pad_top, pad_bottom + htailpad };
            padding_w = { pad_left, pad_right + wtailpad };
        }
        if (pad_mode == 1) // valid padding
        {
            padding_h = { pad_top, pad_bottom };
            padding_w = { pad_left, pad_right };
        }
        if (pad_mode == 2) // tensorflow padding=SAME or onnx padding=SAME_UPPER
        {
            const int w = in_shape[2];
            const int h = in_shape[1];
            const int wpad = kernel_w + (w - 1) / stride_w * stride_w - w;
            const int hpad = kernel_h + (h - 1) / stride_h * stride_h - h;

            padding_h = { hpad / 2, hpad - hpad / 2 };
            padding_w = { wpad / 2, wpad - wpad / 2 };
        }
        if (pad_mode == 3) // onnx padding=SAME_LOWER
        {
            const int w = in_shape[2];
            const int h = in_shape[1];
            const int wpad = kernel_w + (w - 1) / stride_w * stride_w - w;
            const int hpad = kernel_h + (h - 1) / stride_h * stride_h - h;

            padding_h = { hpad - hpad / 2, hpad / 2 };
            padding_w = { wpad - wpad / 2, wpad / 2 };
        }
    }

    if (!pooling_node)
    {
        pooling_node = graph_.emplace<reduce_window2d>(reduce_type, rshape_in_shape, init_value, kernel_h, kernel_w, padding_h, padding_w, stride_h, stride_w, 1, 1, fused_activation, ceil_mode, avgpool_count_include_pad);
        pooling_node->name(op_name + ".reduce_window2d(Pooling)");
    }

    auto out_shape = pooling_node->output_at(0).shape();
    assert(out_shape[0] == 1);
    shape_t rshape_out_shape(out_shape.begin() + 1, out_shape.end());
    auto out_rshape = graph_.emplace<bitcast>(dt_float32, out_shape, rshape_out_shape);
    out_rshape->name(op_name + ".reshape(Output)");

    pooling_node->input_at(0).connect(in_rshape->output());
    out_rshape->input().connect(pooling_node->output_at(0));
    output_tensors_.emplace(layer.tops[0], &out_rshape->output());
}

File: src/Nncase.Importer/Ncnn/Pooling.cs
// Copyright (c) Canaan Inc. All rights reserved.
// Licensed under the Apache license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Nncase.IR;
using Nncase.IR.Buffers;
using Nncase.IR.F;
using Nncase.IR.NN;

namespace Nncase.Importer.Ncnn;

public partial class NcnnImporter
{
    private Expr VisitPooling(NcnnLayer layer)
    {
        var poolingType = layer.ParamDict.Get(0, 0);
        var kernelW = layer.ParamDict.Get(1, 0);
        var kernelH = layer.ParamDict.Get(11, kernelW);
        var strideW = layer.ParamDict.Get(2, 1);
        var strideH = layer.ParamDict.Get(12, strideW);
        var padLeft = layer.ParamDict.Get(3, 0);
        var padRight = layer.ParamDict.Get(14, padLeft);
        var padTop = layer.ParamDict.Get(13, padLeft);
        var padBottom = layer.ParamDict.Get(15, padTop);
        var globalPooling = layer.ParamDict.Get(4, 0) != 0;
        var padMode = layer.ParamDict.Get(5, 0);
        var avgpoolCountIncludePad = layer.ParamDict.Get(6, 0) != 0;
        var adaptivePooling = layer.ParamDict.Get(7, 0) != 0;
        var outW = layer.ParamDict.Get(8, 0);
        var outH = layer.ParamDict.Get(18, outW);

        Expr pooling;
        var input = CHWToNCHW(GetInputExprs(layer, 0));
        (var reduceOp, var initValue) = poolingType switch
        {
            0 => (ReduceOp.Max, float.NegativeInfinity),
            1 => (ReduceOp.Mean, 0f),
            _ => throw new NotSupportedException($"Unsupported pooling type: {poolingType}."),
        };
        var filter = Tensor.From(new[] { kernelH, kernelW }, new[] { 2 });
        var stride = Tensor.From(new[] { strideH, strideW }, new[] { 2 });
        var dilation = Tensor.FromScalar(0, new[] { 2, 2 });

        if (globalPooling)
        {
            pooling = Tensors.Reduce(reduceOp, input, new[] { 2, 3 }, initValue, false);
        }
        else if (adaptivePooling)
        {
            var padding = Tensor.FromScalar(0, new[] { 2, 2 });
            var inShape = Tensors.ShapeOf(input);
            var w = inShape[3];
            var h = inShape[2];
            var kernelExtentH = h - outH + 1;
            var kernelExtentW = w - outW + 1;
            var adaptiveFilter = Tensors.Stack(new IR.Tuple(kernelExtentH, kernelExtentW), 0);
            var adaptiveStride = Tensor.FromScalar(1, 2);

            pooling = NN.ReduceWindow2D(reduceOp, input, initValue, adaptiveFilter, adaptiveStride, padding, dilation, false, avgpoolCountIncludePad);
        }
        else
        {
            Expr[] paddingH;
            Expr[] paddingW;

            if (padMode == 1)
            {
                // valid padding
                paddingH = new Expr[] { padTop, padBottom };
                paddingW = new Expr[] { padLeft, padRight };
            }
            else
            {
                var inShape = Tensors.ShapeOf(input);
                var w = inShape[3];
                var h = inShape[2];

                if (padMode == 0)
                {
                    // full padding
                    var tailW = (w + padLeft + padRight - kernelW) % strideW;
                    var tailH = (h + padTop + padBottom - kernelH) % strideH;

                    var tailPadW = IR.F.Math.Select(IR.F.Math.Equal(tailW, 0), 0, tailW);
                    var tailPadH = IR.F.Math.Select(IR.F.Math.Equal(tailH, 0), 0, tailH);

                    paddingH = new Expr[] { padTop, padBottom + tailPadH };
                    paddingW = new Expr[] { padLeft, padRight + tailPadW };
                }
                else if (padMode is 2 or 3)
                {
                    // valid padding
                    var padH = kernelH + ((h - 1) / strideH * strideH) - h;
                    var padW = kernelW + ((w - 1) / strideW * strideW) - w;

                    if (padMode == 2)
                    {
                        // tensorflow padding=SAME or onnx padding=SAME_UPPER
                        paddingH = new Expr[] { padH / 2, padH - (padH / 2) };
                        paddingW = new Expr[] { padW / 2, padW - (padW / 2) };
                    }
                    else
                    {
                        // onnx padding=SAME_LOWER
                        paddingH = new Expr[] { padH - (padH / 2), padH / 2 };
                        paddingW = new Expr[] { padW - (padW / 2), padW / 2 };
                    }
                }
                else
                {
                    throw new NotSupportedException($"Unsupported pad mode: {padMode}.");
                }
            }

            var padding = Util.ConcatPadding(paddingH, paddingW);
            pooling = NN.ReduceWindow2D(reduceOp, input, initValue, filter, stride, padding, dilation, false, avgpoolCountIncludePad);
        }

        var output = NCHWToCHW(pooling);
        return output;
    }
}
